service-registry/registry-zero-config/src/main/java/org/apache/servicecomb/zeroconfig/client/ZeroConfigClient.java

```diff
@@ -0,0 +1,301 @@
+import java.net.InetAddress;
+import java.net.MulticastSocket;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.lang.StringUtils;
+import org.apache.servicecomb.registry.api.registry.FindInstancesResponse;
+import org.apache.servicecomb.registry.api.registry.Microservice;
+import org.apache.servicecomb.registry.api.registry.MicroserviceInstance;
+import org.apache.servicecomb.registry.api.registry.MicroserviceInstances;
+import org.apache.servicecomb.registry.version.Version;
+import org.apache.servicecomb.registry.version.VersionRule;
+import org.apache.servicecomb.registry.version.VersionRuleUtils;
+import org.apache.servicecomb.registry.version.VersionUtils;
+import org.apache.servicecomb.zeroconfig.ZeroConfigRegistration;
+import org.apache.servicecomb.zeroconfig.server.ServerMicroserviceInstance;
+import org.apache.servicecomb.zeroconfig.server.ServerUtil;
+import org.apache.servicecomb.zeroconfig.server.ZeroConfigRegistryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.client.RestTemplate;
+
+import static org.apache.servicecomb.zeroconfig.ZeroConfigRegistryConstants.*;
+
+public class ZeroConfigClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ZeroConfigClient.class);
+
+  public static ZeroConfigClient INSTANCE = buildZeroConfigClient();
+
+  // Constructor Parameters
+  private ZeroConfigRegistryService zeroConfigRegistryService;
+  private RestTemplate restTemplate;
+  private MulticastSocket multicastSocket;
+
+  // Constructor
+
+  private ZeroConfigClient(ZeroConfigRegistryService zeroConfigRegistryService,
+      MulticastSocket multicastSocket, RestTemplate restTemplate) {
+    this.zeroConfigRegistryService = zeroConfigRegistryService;
+    this.restTemplate = restTemplate;
+    this.multicastSocket = multicastSocket;
+  }
+
+  @VisibleForTesting
+  public ZeroConfigClient initZeroConfigClientWithMocked(
+      ZeroConfigRegistryService zeroConfigRegistryService,
+      MulticastSocket multicastSocket, RestTemplate restTemplate) {
+    this.zeroConfigRegistryService = zeroConfigRegistryService;
+    this.multicastSocket = multicastSocket;
+    this.restTemplate = restTemplate;
+    return this;
+  }
+
+  // builder method
+  private static ZeroConfigClient buildZeroConfigClient() {
+    MulticastSocket multicastSocket = null;
+    try {
+      multicastSocket = new MulticastSocket();
+      multicastSocket.setLoopbackMode(false);
+    } catch (IOException e) {
+      // since we support multiple registries, not blocking other type of registries
+      LOGGER.error("Failed to create MulticastSocket object in Zero-Config mode", e);
+      //throw new ServiceCombException("Failed to create MulticastSocket object", e);
+    }
+    return new ZeroConfigClient(new ZeroConfigRegistryService(), multicastSocket,
+        new RestTemplate());
```

Comment:
Can not use RestTemlate here bacause it is can only invoke simple services that do not deploy any security features, e.g. when TLS is eanbled.

Reply:
will think about perfect soution in next PR